### PR TITLE
Fetch data from NOAA API + send to CODAP table (PT-186719585)

### PR DESCRIPTION
**Pull request by lublagg**

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -38,6 +43,107 @@ export const App = () => {
...
+
+  const fetchErrorHandler = (message: string, resultText: string) => {
+    if (resultText && resultText.length && (resultText[0] === "<")) {
+      try {
+        let xmlDoc = new DOMParser().parseFromString(resultText, "text/xml");
+        message = xmlDoc.getElementsByTagName("userMessage")[0].innerHTML;
+        message += "(" + xmlDoc.getElementsByTagName(
+            "developerMessage")[0].innerHTML + ")";
+      } catch (e: any) {
+        console.log("Error parsing XML: " + e);
```
Comment (eireland) [inline src/components/App.tsx]: Maybe add an linting exception here so we ddon't keep getting the warning

File: src/components/App.tsx [hunk 2]
```diff
@@ -38,6 +43,107 @@ export const App = () => {
...
+          startDate,
+          endDate,
+          frequency: selectedFrequency,
+          weatherStation,
+          attributes,
+          stationTimezoneOffset
+        });
+        try {
+          const tRequest = new Request(tURL);
+          console.log("tRequest", tRequest);
```
Comment (eireland) [inline src/components/App.tsx]: left over console.log?

File: src/components/App.tsx [hunk 3]
```diff
@@ -38,6 +43,107 @@ export const App = () => {
...
+          const tResult = await fetch(tRequest, {mode: "cors"});
+          console.log("fetch result: " + JSON.stringify(tResult));
```
Comment (eireland) [inline src/components/App.tsx]: left over console.log?

File: src/components/App.tsx [hunk 4]
```diff
@@ -38,6 +43,107 @@ export const App = () => {
...
+          setIsFetching(true);
+          if (tResult.ok) {
+            const theJSON = await tResult.json();
+            console.log("theJSON", theJSON);
```
Comment (eireland) [inline src/components/App.tsx]: left over console.log?

File: src/hooks/use-codap-api.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  };
+
+  const arrayify = (value: any) => {
+    return Array.isArray(value) ? value : [value];
+  };
+
+  const createNOAAItems = async (dataRecords: any, dataTypes: IDataType[]) => {
+    await updateWeatherDataset(dataTypes);
+    const items = arrayify(dataRecords);
+    console.log("noaa-cdo ... createNOAAItems with " + dataRecords.length + " case(s)");
```
Comment (eireland) [inline src/hooks/use-codap-api.tsx:149]: left over console.log?

File: src/utils/codapHelpers.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+       && componentsResult.success
+       && componentsResult.values.find(function (component: IMapComponent) {
+           return component.type === "map";
+       });
+};
+
+const createStationsDataset = async (stations: IWeatherStation[]) => {
+  let result = await createDataContext(kStationsDatasetName);
+  if (!result.success) {
+      console.log(`Dataset, "${kStationsDatasetName}", creation failed`);
```
Comment (eireland) [inline src/utils/codapHelpers.ts]: Add a linting exception so we don't get warning

File: src/utils/codapHelpers.ts [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+      return;
+  }
+
+  result = await createNewCollection(kStationsDatasetName, kStationsCollectionName, kWeatherStationCollectionAttrs);
+  if (!result.success) {
+    console.log(`Collection, "${kStationsCollectionName}", creation failed`);
```
Comment (eireland) [inline src/utils/codapHelpers.ts]: Add a linting exception so we don't get warning

File: src/utils/codapHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+};
+
+const arrayify = (value: any) => {
+  return Array.isArray(value) ? value : [value];
+};
+
+const createNOAAItem = async (dsName: string, iValues: any, dataTypes: IDataType[], unitSystem: IUnits) => {
+   await updateWeatherDataset(dataTypes, unitSystem);
+   const items = arrayify(iValues);
+   console.log("noaa-cdo ... createNOAAItems with " + iValues.length + " case(s)");
```
Comment (eireland) [inline src/utils/codapHelpers.ts]: left over console.log?

File: src/utils/codapHelpers.ts (hunk 7 above)
Comment (kswenson) [reply src/utils/codapHelpers.ts]: Or make it a `console.warn` and configure the linter to allow warnings.

File: src/utils/noaaApiHelper.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+          dataTypeName = dataType.name;
+          out[dataTypeName] = decodeData(dataTypeName, value, database);
+        });
+    }
+  });
+  return out;
+};
+
+export const getSelectedStations = (database: string, weatherStation: IWeatherStation) => {
+  let id = database === "global-hourly" ? "isdID": "ghcndID" as keyof IWeatherStation;
```
Comment (eireland) [inline src/utils/noaaApiHelper.ts]:
```suggestion
  let id = database === "global-hourly" ? "isdID" : "ghcndID" as keyof IWeatherStation;
```

File: src/utils/noaaApiHelper.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+      eDate = dayjs(endDate).subtract(stationTimezoneOffset, "hour").add(1, "day");
+  }
+  const startDateString = dayjs(sDate).format(format);
+  const endDateString = dayjs(eDate).format(format);
+  const tDatasetIDClause = `dataset=${database}`;
+  const tStationIDClause = `stations=${getSelectedStations(database, weatherStation).join()}`;
+  const dataTypes = attributes.map(function (attrName) {
+      return dataTypeStore.findByName(attrName)?.sourceName;
+  }); // to-do: update for when attributes are objects
+  console.log("dataTypes", dataTypes);
```
Comment (eireland) [inline src/utils/noaaApiHelper.ts]: left over console.log?

File: src/utils/noaaApiHelper.ts [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  const tDataTypeIDClause = `dataTypes=${dataTypes.join()}`;
+  const tStartDateClause = `startDate=${startDateString}`;
+  const tEndDateClause = `endDate=${endDateString}`;
+  const tUnitClause = `units=metric`;
+  const tFormatClause = "format=json";
+
+  let tURL = [nceiBaseURL, [tDatasetIDClause, tStationIDClause, tStartDateClause, tEndDateClause, tFormatClause, tDataTypeIDClause, tUnitClause].join(
+      "&")].join("?");
+  console.log(`Fetching: ${tURL}`);
```
Comment (eireland) [inline src/utils/noaaApiHelper.ts:117]: add linting exception so we don't get the warning?

**Review by eireland — APPROVED** [review]
There's a bunch of console.logs that should be either excepted or deleted.
Otherwise, looks good.

**Review by eireland — CHANGES_REQUESTED** [review]
There's a bunch of console.logs that either needs to be excepted or deleted

**Review by eireland — APPROVED** [review]
👍 LGTM

**Pull request merged**